#340-created-prisma-tools-doc-and-updated-onboarding (#401)
Northeastern-Electric-Racing/PM-Dashboard-v2: docs/PrismaMigrationTools.md

# Prisma Migration Tools

## `npm prisma:seed`

### What it does

> This command seeds the database, which means populating the database with initial data it requires to run. It can be dummy data, basic user accounts, a default language, etc.

Reviewer: One sentence, one line. This makes it easier for editing in the future.

### Why to use it

> While developing the application, we need some data to be present. Running this command populates the database with the data it needs.

Reviewer: I think we can be more specific. It needs data in order to have anything show up on the front end to interact with. It will populate the database with a basic set of data to use for development purposes.

For more detailed information about [prisma seed](https://www.prisma.io/docs/guides/database/seed-database), click the link.

## `npm prisma:reset`

### What it does

> This command drops the database if possible or performs a soft reset if the database cannot be deleted. If the database is dropped, it creates a new database with the same name and runs all migrations and seed scripts.

Reviewer: Say a little more about how because it runs seeding, then if this is successful you don't need to run the seed command again. Also mention that you use reset if you've sufficiently modified the data in the front end and want to go back to a clean slate with data populated.

Reviewer: Oops maybe not all in that line, but in general

### Why to use it

If you need to undo manual changes or if there are migration history conflicts, you should run this command. This command is a development only tool and should not be used on the production database.

For more detailed information about [prisma reset](https://www.prisma.io/docs/concepts/components/prisma-migrate), click the link.

## `npm prisma:migrate`

### What it does

> This command enables you to keep your database schema and prisma schema in sync with each other.

Reviewer: Probably specify that it is used for updating the database schema based on changes you make to the prisma schema, it's not bidirectional I don't think

### Why to use it

> After you make a change to the prisma schema, you should run this function to update the database schema and create a record of the changes you made (which are shown in the `migrations` folder). This command is a development only tool and should not be used on the production database.

Reviewer: I believe we do use it on the production database, but production database usage of these tools isn't really the main point of this doc. Also I should note that the migrations folder are not an after-the-fact record but instead a set of SQL files generated in order to run on the database and execute the changes

For more detailed information about [prisma migrate](https://www.prisma.io/docs/concepts/components/prisma-migrate), click the link.

## `npm prisma:generate`

### What it does

> This command generates and initializes the prisma client. The prisma client is an auto-generated library that allows us to access data in our application.

Reviewer: One sentence, one line pls

### Why to use it

> After changes are made to the prisma schema, you need to run this function to update the prisma client.

Reviewer: You should also note that I believe npm install will automatically run this script via post install I think (double check and see) so I typically just tell people to run npm install since npm install can potentially solve other update drift issues with the utils package

# Prisma Migration Tools

## `npm prisma:seed`

### What it does

This command seeds the database, which means populating the database with initial data it requires to run.
It can be dummy data, basic user accounts, a default language, etc.

### Why to use it

> While developing the application, we need some data to be present in order for something to appear on the front end to interact with. Running this command populates the database with a basic set of data for development purposes.

Reviewer: one sentence, one line

For more detailed information about [prisma seed](https://www.prisma.io/docs/guides/database/seed-database), click the link.

## `npm prisma:reset`

### What it does

This command drops the database if possible or performs a soft reset if the database cannot be deleted.
If the database is dropped, it creates a new database with the same name and runs all migrations and seed scripts.
If `npm prisma:reset` runs without any errors, then the `npm prisma:seed` function is auto-invoked, meaning you do not need to run the `npm prisma:seed` command again.

### Why to use it

> If you need to undo manual changes or if there are migration history conflicts, you should run this command.

Reviewer: what does "undo manual changes" mean? I think it is too vague.

> You use this command to also restore the data to its original state if you sufficiently modified the data in the front end.

Reviewer: ```suggestion
You use this command to also restore the data to its original state if you sufficiently modified the data via the front end.
```

For more detailed information about [prisma reset](https://www.prisma.io/docs/concepts/components/prisma-migrate), click the link.

> ## `npm prisma:migrate`

Reviewer: don't you also need to describe how to use this one because this one requires naming the migration?

### What it does

This command enables you to keep your database schema in sync with the prisma schema.

### Why to use it

After you make a change to the prisma schema, you should run this function to update the database schema.

> Each time you run this command, it generates an SQL file that is added to the `migrations` folder.

Reviewer: actually it only creates a new SQL file if there's changes to the prisma schema that haven't already been turned into SQL. and you only need to name the newly created migration if SQL file(s) are created. if no new migration is created, this command simply runs the existing SQL files against the database to apply the existing migrations.

## `npm prisma:seed`

### What it does

This command seeds the database, which means populating the database with initial data it requires to run.
It can be dummy data, basic user accounts, a default language, etc.

### Why to use it

While developing the application, we need some data to be present in order for something to appear on the front end to interact with.
Running this command populates the database with a basic set of data for development purposes.

For more detailed information about [prisma seed](https://www.prisma.io/docs/guides/database/seed-database), click the link.

## `npm prisma:reset`

### What it does

This command drops the database if possible or performs a soft reset if the database cannot be deleted.
If the database is dropped, it creates a new database with the same name and runs all migrations and seed scripts.
If `npm prisma:reset` runs without any errors, then the `npm prisma:seed` function is auto-invoked.
This means you do not need to run the `npm prisma:seed` command again.

### Why to use it

If you need to undo changes you made to the database schema or if there are migration history conflicts, you should run this command.

> You use this command to also restore the data to its original state if you sufficiently modified the data via the front end.

Reviewer: let's reverse these two lines. restoring the data is probably the primary reason for devs running this command, and the other sentence describes the potential alternative reasons. what do you think?

For more detailed information about [prisma reset](https://www.prisma.io/docs/concepts/components/prisma-migrate), click the link.

## `npm prisma:migrate`

### What it does

This command enables you to keep your database schema in sync with the prisma schema.

### Why to use it

After you make a change to the prisma schema, you should run this function to update the database schema.

> There will be a `name` field that the terminal will ask you to enter for the migration.

Reviewer: specifically to create the new migration associated with the changes made to the prisma schema.

> Enter a concise and appropriate name that reflects the changes you made to the prisma schema.

Reviewer: love this

Each time you run this command, if there are new changes, a SQL file will be generated that is added to the `migrations` folder.

> Otherwise, `npm prisma:migrate` will run the existing SQL files in the database schema and not require a `name` field to be entered.

Reviewer: perfect

For more detailed information about [prisma migrate](https://www.prisma.io/docs/concepts/components/prisma-migrate), click the link.

## `npm prisma:generate`

### What it does

This command generates and initializes the prisma client.
The prisma client is an auto-generated library that allows us to access data in our application.

> Note: The command `npm install` will automatically run `npm prisma:generate`.

Reviewer: do you think we should add a note about why we set it up this way? I think it's basically to ensure the prisma client is always kept up to date, given that we've also tied the utils package re-generation to npm install. now that I think about it, perhaps these two repository generation items should be de-coupled from npm install since npm install typically fetches remote dependent packages, something we don't necessarily want to do so often. hmmm.

Reviewer 2: I think having the two items coupled with npm install is fine, as npm install is more of a 'catch-all', and the first thing we do when something is not going right with the coding. Instead, we might want to add documentation that explains why and when to run the utils package re-generation and prisma generate so that people can run those separately rather than through `npm install`.

Reviewer 2: Or maybe create a different function that runs both `npm prisma generate` and the utils package regeneration so that if that's all that the user has to run, they could just run that rather than through `npm install`.

Reviewer: Yep! My thinking exactly: make a separate npm script in the package.json to run just the prisma generate and utils package regeneration scripts so that devs can run that instead of always running npm install. And yes, documentation to explain those would be really good too. Can you create a separate ticket for this @jinhyunh?

# Prisma Migration Tools

> ## `npm prisma:seed`

Reviewer: ```suggestion
## `npm run prisma:seed`
```

### What it does

This command seeds the database, which means populating the database with initial data it requires to run.

> It can be dummy data, basic user accounts, a default language, etc.

Reviewer: ```suggestion
It can be dummy data, basic user accounts, a default language, example data in various states, etc.
```

### Why to use it

While developing the application, we need some data to be present in order for something to appear on the front end to interact with.
Running this command populates the database with a basic set of data for development purposes.

For more detailed information about [prisma seed](https://www.prisma.io/docs/guides/database/seed-database), click the link.

> ## `npm prisma:reset`

Reviewer: ```suggestion
## `npm run prisma:reset`
```

### What it does

This command drops the database if possible or performs a soft reset if the database cannot be deleted.
If the database is dropped, it creates a new database with the same name and runs all migrations and seed scripts.

> If `npm prisma:reset` runs without any errors, then the `npm prisma:seed` function is auto-invoked.

Reviewer: ```suggestion
If `npm run prisma:reset` runs without any errors, then the `npm run prisma:seed` function is auto-invoked.
```

> This means you do not need to run the `npm prisma:seed` command again.

Reviewer: ```suggestion
This means you do not need to run the `npm run prisma:seed` command again.
```

### Why to use it

You use this command to restore the data to its original state if you sufficiently modified the data via the front end.
If you need to undo changes you made to the database schema or if there are migration history conflicts, you can also run this command.

For more detailed information about [prisma reset](https://www.prisma.io/docs/concepts/components/prisma-migrate), click the link.

> ## `npm prisma:migrate`

Reviewer: ```suggestion
## `npm run prisma:migrate`
```

### What it does

This command enables you to keep your database schema in sync with the prisma schema.

### Why to use it

After you make a change to the prisma schema, you should run this function to update the database schema.
The new migration created is associated with the changes made to the prisma schema.

There will be a `name` field that the terminal will ask you to enter for the migration.
Enter a concise and appropriate name that reflects the changes you made to the prisma schema.

Each time you run this command, if there are new changes, a SQL file will be generated that is added to the `migrations` folder.

> Otherwise, `npm prisma:migrate` will run the existing SQL files in the database schema and not require a `name` field to be entered.

Reviewer: ```suggestion
Otherwise, `npm run prisma:migrate` will apply the existing migration SQL files onto the database and not require a `name` field to be entered.
```

For more detailed information about [prisma migrate](https://www.prisma.io/docs/concepts/components/prisma-migrate), click the link.

> ## `npm prisma:generate`

Reviewer: ```suggestion
## `npm run prisma:generate`
```

### What it does

This command generates and initializes the prisma client.
The prisma client is an auto-generated library that allows us to access data in our application.

> Note: The command `npm install` will automatically run `npm prisma:generate`.

Reviewer: ```suggestion
Note: The command `npm install` will automatically run `npm run prisma:generate`.
```
